### PR TITLE
WIP: Working from S3 to events

### PR DESCRIPTION
**Pull request by tcoupland**

### CODE REVIEW

File: src/kixi/event_replay.clj [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+  [barrays]
+  (when (first barrays)
+    (let [[nippy-seq size] (take-nippy-seq barrays)]
+      (cons nippy-seq
+            (lazy-seq 
+             (combine-nippy-header-sequences
+              (drop size
+                    barrays)))))))
+
+(defn object-summery->input-stream
```
Comment (otfrom) [inline src/kixi/event_replay.clj]: Do you mean summAry?

File: doc/intro.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# Introduction to kixi.event-replay
```
Comment (otfrom) [inline doc/intro.md:1]: remove

File: README.md [hunk 3]
```diff
@@ -1,2 +1,14 @@
 # kixi.event-replay
-Event backup replayer
+
```
Comment (otfrom) [inline README.md:2]: update w/correct text

File: src/kixi/event_replay.clj [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+  (transduce execute-replay 
+             aggregate-events
+             (hour-sequence)))
+
+(defn exit
+  [status msg]
+  (println msg)
+  (System/exit status))
+
+(defn -main
```
Comment (otfrom) [inline src/kixi/event_replay.clj:170]: always via the config and not dates passed in?

File: src/kixi/event_replay.clj [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+  [x]
+  (prn x)
+  x)
+
+(def execute-replay
+  (comp (mapcat hour->s3-object-summaries)
+        (take 2)
+        (map object-summary->input-stream)
+        input-stream->events
+        (map prn-t)
```
Comment (otfrom) [inline src/kixi/event_replay.clj:150]: Do as a proper logger?

File: src/kixi/event_replay.clj [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+  :done)
+
+(defn prn-t
+  [x]
+  (prn x)
+  x)
+
+(def execute-replay
+  (comp (mapcat hour->s3-object-summaries)
+        (take 2)
```
Comment (otfrom) [inline src/kixi/event_replay.clj:147]: Limit just for testing?

File: src/kixi/event_replay.clj [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+             (when (:next-marker list-objects-res)
+               (hour->s3-object-summaries s3-base-dir prefix (:next-marker list-objects-res)))))))
+
+(defn object-summary->input-stream
+  [s3-object-summary]
+  (let [{:keys [s3-base-dir
+                target-stream]} @config
+        s3-object (s3/get-object :bucket-name s3-base-dir
+                                 :key (:key s3-object-summary))]
+    (:object-content s3-object)))
```
Comment (otfrom) [inline src/kixi/event_replay.clj:111]: Should this not be `:input-stream`?

File: src/kixi/event_replay.clj [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+        start (parse-datahour start-datehour)
+        end (parse-datahour end-datehour)]    
+    (p/periodic-seq start
+                    (t/plus end
+                            one-hour)
+                    one-hour)))
+
+(defn hour->s3-prefix
+  [hour]
+  (->> [(t/year hour)
```
Comment (otfrom) [inline src/kixi/event_replay.clj:77]: can't you use `clj-time.format/unparse` here?

File: src/kixi/event_replay/inputstream_to_events.clj [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+  (:require [gloss.io :as gloss.io]
+            [taoensso.nippy :as nippy])
+  (:import [java.nio ByteBuffer]
+           [java.io InputStream]))
+
+(def line-break 10)
+
+(def decode-buffer-size (int 1024))
+
+(defn input-stream->lines
```
Comment (otfrom) [inline src/kixi/event_replay/inputstream_to_events.clj:11]: Worth looking at `lines-reduceable` here: https://tech.grammarly.com/blog/building-etl-pipelines-with-clojure ?

**Review by otfrom — APPROVED** [review]
Some things I'm a bit confused by and some for you to think about (like the reducible resource), but looks like the right direction.

File: src/kixi/event_replay.clj (hunk 8 above)
Comment (tcoupland) [reply src/kixi/event_replay.clj:77]: Could create and slash delimited format I guess, but wldnt gain too much.

File: src/kixi/event_replay.clj (hunk 7 above)
Comment (tcoupland) [reply src/kixi/event_replay.clj:111]: Sadly not, :object-content is an input stream of the objects contents

File: src/kixi/event_replay.clj (hunk 6 above)
Comment (tcoupland) [reply src/kixi/event_replay.clj:147]: Yep

File: src/kixi/event_replay.clj (hunk 5 above)
Comment (tcoupland) [reply src/kixi/event_replay.clj:150]: Yeah, will need something to tell us the progress

File: src/kixi/event_replay.clj (hunk 4 above)
Comment (tcoupland) [reply src/kixi/event_replay.clj:170]: This is going to be executed via docker, so preferring env vars, as makes the docker file cleaner.

File: src/kixi/event_replay/inputstream_to_events.clj (hunk 9 above)
Comment (tcoupland) [reply src/kixi/event_replay/inputstream_to_events.clj:11]: That is a great blog post! Trouble with any line based stuff is it strings the contents, current stuff keeps it raw bytes until it gets to nippy thaw. Will check that out though, the line breaker is a bit messy.

**Pull request merged**